##This scripts were download from HuaZou github. It is very good script. I used it to download TCGA data. Another benifit to use  this scripts to integrate with Dr. Reppin's lab machine learning model https://github.com/ruppinlab/tcga-microbiome-prediction 
#I modified the scripts to adapt the individual dataset, either for RNA, metagenomics and epigenomics data
#The related easybuild environment file has been list in easyguild folder.  


## Download Clinical information and Omic-data profile from TCGA project

### Introduction

TCGA project had produced multiple omics data in Cancer research, and the public data could be assessed by R program. This pipeline aims to acquire the clinical information and omic-data profile and dissect the information into analysis model.


### How it works

There are four scripts which separated the pipeline into four parts

1. **01.Download.R**: download clinical information and save omics data into SummariedExperimentData

2. **02.Prerpocess_clinical.R**: Process the clinical information

3. **03.Prerpocess_omics.R**: Process the omics data profile

4. **04.ExpressionSet.R**: Convert Phenotype from clincal information and Profile into ExpressionSet


### Materials

Downloading the reference files from GDC for the Copy number variation to get the CNV markers is necessary, and the file named **util/snp6.na35.remap.hg38.subset.txt.gz**. In the other hand, transforming the *ensembl geneid* into *gene symbol* via the **util/human_gene_all.tsv** file which generated by biomart R packages would be convenient for subsequent study. 


### Command line 

Since these R scripts are used in linux system envirnment, the following command would display that how to excute the pipeline with the TCGA-PAAD example. 

```bash
# step1: Download the multiple omics data from TCGA
Rscript 01.Download.R -t TCGA-PAAD -o CNV
Rscript 01.Download.R -t TCGA-PAAD -o mRNA
Rscript 01.Download.R -t TCGA-PAAD -o miRNA
Rscript 01.Download.R -t TCGA-PAAD -o DNA_Methylation

# step2: Curate the origin Rawdata including clinical information and omics data  
Rscript 02.Prerpocess_clinical.R -p Clinical/TCGA-PAAD_clinical_origin.csv -t TCGA-PAAD

Rscript 03.Prerpocess_omics.R -e Omics/TCGA-PAAD_CNV.RDS -t CNV -g util/snp6.na35.remap.hg38.subset.txt.gz -p TCGA-PAAD-post
Rscript 03.Prerpocess_omics.R -e Omics/TCGA-PAAD_mRNA.RDS -t mRNA -g util/human_gene_all.tsv -p TCGA-PAAD-post
Rscript 03.Prerpocess_omics.R -e Omics/TCGA-PAAD_miRNA.RDS -t miRNA -p TCGA-PAAD-post
Rscript 03.Prerpocess_omics.R -e Omics/TCGA-PAAD_DNA_Methylation.RDS -t DNA_Methylation -p TCGA-PAAD-post

# step3: Convert clinical information and Gene expression data into ExpressionSet for further analysis
Rscript 04.ExpressionSet.R -p Clinical/TCGA-PAAD-post_clinical.csv -e Clean/TCGA-PAAD-post_mRNA_profile.tsv -t mRNA -o TCGA-PAAD-post
Rscript 04.ExpressionSet.R -p Clinical/TCGA-PAAD-post_clinical.csv -e Clean/TCGA-PAAD-post_miRNA_PKM.tsv -t miRNA -o TCGA-PAAD-post-PKM
Rscript 04.ExpressionSet.R -p Clinical/TCGA-PAAD-post_clinical.csv -e Clean/TCGA-PAAD-post_miRNA_count.tsv -t miRNA -o TCGA-PAAD-post-count
```

**Notes: Since accessing to TCGA website via the GDC API, using R TCGAbiolinks package does it would be rejected. To address this problem is try more and more times until the download is successful.**

### Summary

The clinical information contains all the patients' phenotype but it doesn't show any Normal samples' information derived from omics data profile. Therefore, the extra information about normal samples would be obtains from the SummariedExperimentData. In the other hand, to distinct the tumor and normal samples is also be cautious, the Number named "01" at the position from 14 to 15 characters of SummarizedExperiment object's rownames is "tumor" and the others are "normal".


### Files' Structure

The following tree of file structure displays the whole final results of the downloading, preprocessing and integration of the TCGA-PAAD project from TCGA website through R program.

The final file for further analysis are named with *RDS* or *post* suffix in the Clean and Clinical diretories.

```bash
TCGA_download_pipeline
├── 01.Download.R
├── 02.Prerpocess_clinical.R
├── 03.Prerpocess_omics.R
├── 04.ExpressionSet.R
├── CNV
│   └── TCGA-PAAD
│       └── harmonized
├── Clean
│   ├── TCGA-PAAD-post-PKM_miRNA_ExprSet_clinical.RDS
│   ├── TCGA-PAAD-post-count_miRNA_ExprSet_clinical.RDS
│   ├── TCGA-PAAD-post_CNV_tumor.tsv
│   ├── TCGA-PAAD-post_CNV_untumor.tsv
│   ├── TCGA-PAAD-post_DNA_Methylation_ExprSet.RDS
│   ├── TCGA-PAAD-post_DNA_Methylation_clinical.csv
│   ├── TCGA-PAAD-post_DNA_Methylation_profile.tsv
│   ├── TCGA-PAAD-post_mRNA_ExprSet.RDS
│   ├── TCGA-PAAD-post_mRNA_ExprSet_clinical.RDS
│   ├── TCGA-PAAD-post_mRNA_clinical.csv
│   ├── TCGA-PAAD-post_mRNA_profile.tsv
│   ├── TCGA-PAAD-post_miRNA_PKM.tsv
│   ├── TCGA-PAAD-post_miRNA_count.tsv
│   └── snp6.na35.remap.hg38.tsv
├── Clinical
│   ├── TCGA-PAAD-post_clinical.csv
│   └── TCGA-PAAD_clinical_origin.csv
├── DNA_Methylation
│   └── TCGA-PAAD
│       └── legacy
├── MANIFEST.txt
├── Omics
│   ├── TCGA-PAAD_CNV.RDS
│   ├── TCGA-PAAD_DNA_Methylation.RDS
│   ├── TCGA-PAAD_mRNA.RDS
│   └── TCGA-PAAD_miRNA.RDS
├── README.md
├── mRNA
│   └── TCGA-PAAD
│       └── harmonized
├── miRNA
│   └── TCGA-PAAD
│       └── harmonized
├── util
│   ├── human_gene_all.tsv
│   └── snp6.na35.remap.hg38.subset.txt.gz
└── work.sh
```


### Reference 

1. [GDC Reference](https://gdc.cancer.gov/about-data/gdc-data-processing/gdc-reference-files)

2. [How to perform CNV analysis](https://www.jianshu.com/p/b8351a86a40b)

